pkgs/applications/virtualization/qemu/encode-video.c

```diff
@@ -0,0 +1,737 @@
+ * or implicitly).
+ */
+static bool convert_endian(void *data, size_t len, uint8_t bpp)
+{
+    const uint16_t endian_test = 1;
+
+    if (*(uint8_t*)&endian_test == 0)
+        return true;
+
+    switch (bpp) {
+        case 2:
+            for (int i = 0; i < len; ++i, data += bpp)
+                *(uint16_t*)data = bswap_16(*(uint16_t*)data);
+            break;
+        case 4:
+            for (int i = 0; i < len; ++i, data += bpp)
+                *(uint32_t*)data = bswap_32(*(uint32_t*)data);
+            break;
+        default:
+            fprintf(stderr, "Unable to handle pixel byte size of %d.\n", bpp);
+            return false;
+    }
+    return true;
+}
+
+static AVFrame *alloc_frame(enum AVPixelFormat pix_fmt, uint32_t width,
+                            uint32_t height)
+{
+    AVFrame *frame;
+
+    if ((frame = av_frame_alloc()) == NULL)
+        return NULL;
+
+    frame->format = pix_fmt;
+    frame->width  = width;
+    frame->height = height;
+
+    if (av_frame_get_buffer(frame, 0) < 0) {
+        fprintf(stderr, "Could not allocate frame data.\n");
+        return NULL;
+    }
+
+    return frame;
+}
+
+/* Parse a switch packet (indicated by an 'S' byte).
+ *
+ * The format is (numbers after colons are bit sizes):
+ *
+ * <<Width:32, Height:32, Format:8, BPP:8>>
+ */
+static struct packet_switch *parse_switch(gzFile ifile)
+{
+    void *buf;
+    struct packet_switch *out;
+    uint8_t tmp_format;
+
+    if ((buf = alloc_read(ifile, 10)) == NULL)
+        return NULL;
+
+    if ((out = malloc(sizeof(struct packet_switch))) == NULL) {
+        fprintf(stderr, "Unable to allocate packet_switch: %s\n",
+                strerror(errno));
+        is_error = true;
+        free(buf);
+        return NULL;
+    }
+
+    memcpy(&out->width,  buf,     4);
+    memcpy(&out->height, buf + 4, 4);
+    memcpy(&tmp_format,  buf + 8, 1);
+    memcpy(&out->bpp,    buf + 9, 1);
+
+    free(buf);
+
+    switch (tmp_format) {
+        case 1: out->format = AV_PIX_FMT_RGB555BE; break;
+        case 2: out->format = AV_PIX_FMT_RGB565BE; break;
+        case 3: out->format = AV_PIX_FMT_0RGB;     break;
+        case 4: out->format = AV_PIX_FMT_RGB0;     break;
+        case 5: out->format = AV_PIX_FMT_BGR0;     break;
+        default:
+            fprintf(stderr, "Unknown pixel format %d in switch directive.\n",
+                    tmp_format);
+            is_error = true;
+            free(out);
+            return NULL;
+    }
+
+    return out;
+}
+
+/* Parse an update packet (indicated by an 'U' byte).
+ *
+ * The format is (numbers after colons are bit sizes):
+ *
+ * <<Op:8, X:32, Y:32, W:32, H:32, Time:64>>
+ *
+ * The actual data length is determined by the given bytes per pixel
+ * and the width and height of the update packet, which is also saved
+ * into the returned struct.
+ */
+static struct packet_update *parse_update(gzFile ifile, uint8_t bpp)
```

Comment:
wouldn't parse_update and parse_switch be safer (because you don't need to do resource management yourself) and faster (because we need less heap calls) if you either returned such a struct by value or took the struct buffer as a pointer output parameter for this function?

Reply:
Note btw. that this is only the video encoder for the intermediate format, which is *not* used at runtime. So while passing the stack-allocated struct here would be faster, it isn't critical. Nevertheless, I'd like to refactor/rewrite the video encoder since there are a few other issues with the output formats.